sharedbuffers/mapped_struct.py

```diff
@@ -1930,52 +1930,75 @@ def __iter__(self):
                     elif not self.bitrep_hi >> i:
                         break
 
-    @cython.locals(seqs=tuple, pfset='proxied_frozenset')
-    def union(self, *seqs):
-        if not seqs:
-            return self
-        elif self.objlist is None and len(seqs) == 1 and type(seqs[0]) is proxied_frozenset:
-            pfset = cython.cast(proxied_frozenset, seqs[0])
+    @cython.ccall
+    @cython.locals(pfset='proxied_frozenset')
+    def _union_2(self, seq):
+        if self.objlist is None and type(seq) is proxied_frozenset:
+            pfset = cython.cast(proxied_frozenset, seq)
             if pfset.objlist is None:
                 return proxied_frozenset(
                     None, self.bitrep_lo | pfset.bitrep_lo, self.bitrep_hi | pfset.bitrep_hi)
+        return frozenset(self).union(seq)
```

Comment:
You'll be iterating `self` twice at least. Once to build the left frozenset, and then to compute the union and build the resulting frozenset.

That should be akin to:

``` python
rv = set(self)
rv.update(seq)
return frozenset(rv)
```

In performance. Why don't you test it?

Same in all other operations. You're paying the construction cost twice, once for the temporary, and once for the result. So it may be OK to go through a regular set first and optimize the operation.

Reply:
Creating a set with fast iteration and performing inplace updates on it does seem to improve performance. What doesn't have as much impact is using the `merge` interface with integers (only 2 to 5% improvement). I would like to know your opinion, but I don't think it's worth it.

Reply:
If it isn't worth it, it isn't worth it. We can revisit it later in any case.

Paste the timeit script and timings if you can though.

Reply:
Here's the script I used:
```python
from sharedbuffers import mapped_struct
from functools import partial
import timeit

def make_obj(start, end):
    base = [x * 2 for x in xrange(start, end)] + [2 ** 30]
    buf = bytearray(1024 * 16)
    mapped_struct.proxied_frozenset.pack_into(frozenset(base), buf, 0)
    return mapped_struct.proxied_frozenset.unpack_from(buf, 0)

if __name__ == '__main__':
    obj1 = make_obj(0, 1000)
    obj2 = make_obj(500, 1500)
```

And the timings:
- Naive: 126.80619231505
- Fast conversion: 110.266254187
- Merge: 105.635071511146